polaris-service/src/main/java/org/apache/polaris/service/persistence/MetadataCacheManager.java

```diff
@@ -0,0 +1,140 @@
+   */
+  public static TableMetadata loadTableMetadata(
+      TableIdentifier tableIdentifier,
+      long maxBytesToCache,
+      PolarisCallContext callContext,
+      PolarisMetaStoreManager metastoreManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView,
+      Supplier<TableMetadata> fallback) {
+    LOGGER.debug(String.format("Loading cached metadata for %s", tableIdentifier));
+    PolarisResolvedPathWrapper resolvedEntities =
+        resolvedEntityView.getResolvedPath(tableIdentifier, PolarisEntitySubType.TABLE);
+    TableLikeEntity tableLikeEntity = TableLikeEntity.of(resolvedEntities.getRawLeafEntity());
+    boolean isCacheValid =
+        tableLikeEntity.getMetadataLocation().equals(tableLikeEntity.getMetadataCacheLocationKey());
+    if (isCacheValid) {
+      LOGGER.debug(String.format("Using cached metadata for %s", tableIdentifier));
+      return TableMetadataParser.fromJson(tableLikeEntity.getMetadataCacheContent());
+    } else {
+      TableMetadata metadata = fallback.get();
+      PolarisMetaStoreManager.EntityResult cacheResult =
+          cacheTableMetadata(
+              tableLikeEntity,
+              metadata,
+              maxBytesToCache,
+              callContext,
+              metastoreManager,
+              resolvedEntityView);
+      if (!cacheResult.isSuccess()) {
+        LOGGER.debug(String.format("Failed to cache metadata for %s", tableIdentifier));
+      }
+      return metadata;
+    }
+  }
+
+  /**
+   * Attempt to add table metadata to the cache
+   *
+   * @return The result of trying to cache the metadata
+   */
+  private static PolarisMetaStoreManager.EntityResult cacheTableMetadata(
+      TableLikeEntity tableLikeEntity,
+      TableMetadata metadata,
+      long maxBytesToCache,
+      PolarisCallContext callContext,
+      PolarisMetaStoreManager metaStoreManager,
+      PolarisResolutionManifestCatalogView resolvedEntityView) {
+    String json = TableMetadataParser.toJson(metadata);
+    // We should not reach this method in this case, but check just in case...
+    if (maxBytesToCache != PolarisConfiguration.METADATA_CACHE_MAX_BYTES_NO_CACHING) {
+      long sizeInBytes = json.getBytes(StandardCharsets.UTF_8).length;
+      if (sizeInBytes > maxBytesToCache) {
+        LOGGER.debug(
+            String.format(
+                "Will not cache metadata for %s; metadata is %d bytes and the limit is %d",
+                tableLikeEntity.getTableIdentifier(), sizeInBytes, maxBytesToCache));
+        return new PolarisMetaStoreManager.EntityResult(
+            PolarisMetaStoreManager.ReturnStatus.SUCCESS, null);
+      } else {
+        LOGGER.debug(
+            String.format("Caching metadata for %s", tableLikeEntity.getTableIdentifier()));
+        TableLikeEntity newTableLikeEntity =
+            new TableLikeEntity.Builder(tableLikeEntity)
+                .setMetadataContent(tableLikeEntity.getMetadataLocation(), json)
```

Comment:
Looks like this is passing in `tableLikeEntity.getMetadataLocation()` instead of `metadata.metadataFileLocation()` anyways, so it's extraneous. Probably what we want to do here is check that `tableLikeEntity.getMetadataLocation().equals(metadata.metadataFileLocation())` and if it mismatches, it means another write concurrently updated the TableLikeEntity already after our original table resolution.

In that case it's probably still fine to return the stale JSON contents as-is, but just skip trying to touch the cache content here (normally we'd expect the write to already have pre-populated the cache content). We could return some return status for "concurrent modification" or similar if we want the callsite to have a nice debug message.

Reply:
It's not necessarily concurrent modification. Imagine a case where caching is enabled, some JSON is cached, and then caching is disabled for a long time where many updates occur. Afterwards, caching is re-enabled. We need to do this check to make sure we don't serve the very-old metadata from the cache.

Reply:
Makes sense. Could you please add a comment somewhere (maybe at the key definition at the top) explaining the relationship to the main `metadata-location` field?